app/scripts/controllers/network/network-controller.ts

```diff
@@ -0,0 +1,1171 @@
+import { strict as assert } from 'assert';
+import EventEmitter from 'events';
+import { ComposedStore, ObservableStore } from '@metamask/obs-store';
+import log from 'loglevel';
+import {
+  createSwappableProxy,
+  createEventEmitterProxy,
+  SwappableProxy,
+} from '@metamask/swappable-obj-proxy';
+import EthQuery from 'eth-query';
+import { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { v4 as uuid } from 'uuid';
+import { Hex, isPlainObject } from '@metamask/utils';
+import { errorCodes } from 'eth-rpc-errors';
+import { SafeEventEmitterProvider } from '@metamask/eth-json-rpc-provider';
+import { PollingBlockTracker } from 'eth-block-tracker';
+import {
+  INFURA_PROVIDER_TYPES,
+  INFURA_BLOCKED_KEY,
+  TEST_NETWORK_TICKER_MAP,
+  CHAIN_IDS,
+  NETWORK_TYPES,
+  BUILT_IN_INFURA_NETWORKS,
+  BuiltInInfuraNetwork,
+  NetworkStatus,
+} from '../../../../shared/constants/network';
+import {
+  isPrefixedFormattedHexString,
+  isSafeChainId,
+} from '../../../../shared/modules/network.utils';
+import {
+  MetaMetricsEventCategory,
+  MetaMetricsEventPayload,
+} from '../../../../shared/constants/metametrics';
+import { isErrorWithMessage } from '../../../../shared/modules/error';
+import {
+  createNetworkClient,
+  NetworkClientType,
+} from './create-network-client';
+
+/**
+ * The name of NetworkController.
+ */
+const name = 'NetworkController';
+
+/**
+ * A block header object that `eth_getBlockByNumber` returns. Note that this
+ * type does not specify all of the properties present within the block header;
+ * within NetworkController, we are only interested in `baseFeePerGas`.
+ */
+type Block = {
+  baseFeePerGas?: unknown;
+};
+
+/**
+ * Encodes both whether or not a provider is configured for an Infura network or
+ * a non-Infura network; if for an Infura network, then which network; and if
+ * for a non-Infura network, then whether the network is being operated locally
+ * or remotely. Primarily used to build the network client and check the
+ * availability of a network.
```

Comment:
Nit: This sentence was confusing. Maybe we could simplify it somehow? The structure of "Both [first thing thing]; if ... then [second thing]; if ... then [second thing]" was hard to follow and doesn't seem grammatically correct.

Reply:
You're right, this was confusing. Hopefully this is better: [84140af](https://github.com/MetaMask/metamask-extension/pull/18358/commits/84140afaf0b2ea0dd70a6bc14fa9cfcbad1c11cb)

Reply:
That looks great! I'm glad I didn't suggest an alternative, this is better than what I would have come up with.